### PR TITLE
Podman rootless fixes

### LINKED ISSUE #670: [Error] unable to start container "...": crun: openat2 `dev/shm`: No such file or directory: OCI runtime attempted to invoke a command that was not found
**Describe the bug**
Unable to run `distrobox enter` on WSL2 when using rootless `podman`.

**To Reproduce**
Install [Fedora 37 on WSL2](https://github.com/WhitewaterFoundry/Fedora-Remix-for-WSL).

Install Podman `sudo dnf install -y podman`

Create container `distrobox create`

Enter container `distrobox enter`

**Expected behavior**

**Logs**
```
$ distrobox enter
Container my-distrobox is not running.
Starting container my-distrobox
run this command to follow along:

 podman logs -f my-distrobox

Error: unable to start container "9c9d2eba2dc93066e02506f60bc4ff4e640b12f9d792442a8e62e15e2ae52025": crun: openat2 `dev/shm`: No such file or directory: OCI runtime attempted to invoke a command that was not found
```

**Desktop (please complete the following information):**
```
$ wsl --status
Default Distribution: Ubuntu-20.04
Default Version: 2

Windows Subsystem for Linux was last updated on 02/03/2023
WSL automatic updates are on.

Kernel version: 5.10.16
```

```
$ distrobox --version
distrobox: 1.4.2.1
```

```
$ podman info
host:
  arch: amd64
  buildahVersion: 1.29.0
  cgroupControllers: []
  cgroupManager: cgroupfs
  cgroupVersion: v1
  conmon:
    package: conmon-2.1.7-2.fc37.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.1.7, commit: '
  cpuUtilization:
    idlePercent: 99.78
    systemPercent: 0.13
    userPercent: 0.09
  cpus: 6
  distribution:
    distribution: fedoraremixforwsl
    version: "37"
  eventLogger: file
  hostname: VD011109
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 165536
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 165536
      size: 65536
  kernel: 5.10.16.3-microsoft-standard-WSL2
  linkmode: dynamic
  logDriver: k8s-file
  memFree: 10473615360
  memTotal: 26905296896
  networkBackend: netavark
  ociRuntime:
    name: crun
    package: crun-1.8.1-1.fc37.x86_64
    path: /usr/bin/crun
    version: |-
      crun version 1.8.1
      commit: f8a096be060b22ccd3d5f3ebe44108517fbf6c30
      rundir: /tmp/podman-run-1000/crun
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +CRIU +LIBKRUN +WASM:wasmedge +YAJL
  os: linux
  remoteSocket:
    path: /tmp/podman-run-1000/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID
    rootless: true
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: false
  serviceIsRemote: false
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: slirp4netns-1.2.0-8.fc37.x86_64
    version: |-
      slirp4netns version 1.2.0
      commit: 656041d45cfca7a4176f6b7eed9e4fe6c11e8383
      libslirp: 4.7.0
      SLIRP_CONFIG_VERSION_MAX: 4
      libseccomp: 2.5.3
  swapFree: 7516192768
  swapTotal: 7516192768
  uptime: 16h 16m 50.00s (Approximately 0.67 days)
plugins:
  authorization: null
  log:
  - k8s-file
  - none
  - passthrough
  - journald
  network:
  - bridge
  - macvlan
  volume:
  - local
registries:
  search:
  - registry.fedoraproject.org
  - registry.access.redhat.com
  - docker.io
  - quay.io
store:
  configFile: /home/mschoep/.config/containers/storage.conf
  containerStore:
    number: 2
    paused: 0
    running: 0
    stopped: 2
  graphDriverName: overlay
  graphOptions: {}
  graphRoot: /home/mschoep/.local/share/containers/storage
  graphRootAllocated: 269490393088
  graphRootUsed: 16969031680
  graphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "false"
  imageCopyTmpDir: /var/tmp
  imageStore:
    number: 2
  runRoot: /tmp/containers-user-1000/containers
  transientStore: false
  volumePath: /home/mschoep/.local/share/containers/storage/volumes
version:
  APIVersion: 4.4.2
  Built: 1677669779
  BuiltTime: Wed Mar  1 12:22:59 2023
  GitCommit: ""
  GoVersion: go1.19.6
  Os: linux
  OsArch: linux/amd64
  Version: 4.4.2
```

```
$ cat /etc/os-release 
NAME="Fedora Remix for WSL"
VERSION="37"
ID=fedoraremixforwsl
ID_LIKE=fedora
VERSION_ID=37
PLATFORM_ID="platform:f37"
PRETTY_NAME="Fedora Remix for WSL"
ANSI_COLOR="0;34"
CPE_NAME="cpe:/o:fedoraproject:fedora:37"
HOME_URL="https://github.com/WhitewaterFoundry/Fedora-Remix-for-WSL"
SUPPORT_URL="https://github.com/WhitewaterFoundry/Fedora-Remix-for-WSL"
BUG_REPORT_URL="https://github.com/WhitewaterFoundry/Fedora-Remix-for-WSL/issues"
PRIVACY_POLICY_URL="https://github.com/WhitewaterFoundry/Fedora-Remix-for-WSL/blob/master/PRIVACY.md"
FEDORA_REMIX_VERSION=37.0.4
```

**Additional context**

When removing the /dev/shm volume in `distrobox-create` it starts up. 

I will create a PR to force disabling shm later. If an alternative is preferred, please add it to this report.

### PR DESCRIPTION
**Pull request by OhMyMndy**
Do not mount `/dev` and use `--ipc host` when using Podman rootless because it is not supported.

Mounting `/dev` and `/dev/shm`:
```
$ distrobox enter
Container my-distrobox is not running.
Starting container my-distrobox
run this command to follow along:

 podman logs -f my-distrobox

Error: unable to start container "9c9d2eba2dc93066e02506f60bc4ff4e640b12f9d792442a8e62e15e2ae52025": crun: openat2 `dev/shm`: No such file or directory: OCI runtime attempted to invoke a command that was not found
```

Using `--ipc host`
```
$ podman run --rm -it --ipc host alpine Error: statfs /dev/mqueue: no such file or directory
```

Fixes #670

### CODE REVIEW

**Comment by 89luca89** [comment]
Thanks @OhMyMndy for the PR!

I'd like to check the implications of removing `rslave` on systems where it is not supported
Generally speaking if the basic functionalities of distrobox (for example, sharing mounts/drives) are affected by this, I prefer to keep the incompatibility and not having half-working situations

**Review by 89luca89 — CHANGES_REQUESTED** [review]

File: distrobox-create [hunk 1]
```diff
@@ -535,7 +583,7 @@ generate_command() {
 	# to /run/shm, instead of the other way around.
 	# Resolve this detecting if /dev/shm is a symlink and mount original
 	# source also in the container.
-	if [ -L "/dev/shm" ]; then
+	if [ -L "/dev/shm" ] && [ "$use_shm" = "1" ]; then
```
[the remark this answers is not in the document]
Comment (OhMyMndy) [reply distrobox-create]: The strange thing is, `/dev/shm` does exist (which is actually `/run/shm`

```
$ ls -lash /dev/shm/
total 0
0 drwxrwxrwt 2 root root  40 Mar 23 11:48 .
0 drwxr-xr-x 8 root root 160 Mar 23 11:48 ..
```

```
$ distrobox enter
Container my-distrobox is not running.
Starting container my-distrobox
run this command to follow along:

 podman logs -f my-distrobox

Error: unable to start container "1c256f802b075da3d5ee6af4a1c7833a5e31bb44d1220d588cdebe768f3499ad": crun: openat2 `dev/shm`: No such file or directory: OCI runtime attempted to invoke a command that was not found
```

[the remark this answers is not in the document]
Comment (89luca89) [reply distrobox-create]: what does `ls -ladH /dev/shm` report?

[the remark this answers is not in the document]
Comment (OhMyMndy) [reply distrobox-create]:
```
$ ls -ladH /dev/shm
drwxrwxrwt 2 root root 60 Mar 23 11:50 /dev/shm

$ ls -ladH /dev/shm/
drwxrwxrwt 2 root root 60 Mar 23 11:50 /dev/shm/
```

[the remark this answers is not in the document]
Comment (OhMyMndy) [reply distrobox-create]: But also, it looks like `libpod` is using `/dev/shm`...

```
$ ls -lash /dev/shm/
total 84K
  0 drwxrwxrwt 2 root    root     60 Mar 23 11:50 .
  0 drwxr-xr-x 9 root    root    180 Mar 23 11:50 ..
84K -rw------- 1 mschoep mschoep 81K Mar 23 11:50 libpod_rootless_lock_1000
```

[the remark this answers is not in the document]
Comment (OhMyMndy) [reply distrobox-create]: https://github.com/microsoft/WSL/issues/8996#issuecomment-1295953046

This is a "solution" to the problem, so it won't fail here on `/dev/shm`, I am not running the Windows Store version of WSL (which is new, and the only version which supports Systemd)...

Let me see if manually executing those `mkdir` and `mount` commands is "working".

[the remark this answers is not in the document]
Comment (OhMyMndy) [reply distrobox-create]: Looks like mounting `/dev` at all causes problems with rootless Podman

```
$ podman run --rm -it -v /dev:/dev alpine
Error: crun: openat2 `dev/shm`: No such file or directory: OCI runtime attempted to invoke a command that was not found
```

[the remark this answers is not in the document]
Comment (OhMyMndy) [reply distrobox-create]: Trying to run Podman rootless while mounting tmpfs on `/dev/shm` also doesn't work:

With mounting `/dev`:
```
$ podman run --rm -it  --mount type=tmpfs,destination=/dev/shm -v /dev:/dev alpine
Error: crun: openat2 `dev/shm`: No such file or directory: OCI runtime attempted to invoke a command that was not found
```

Without mounting `/dev`:
```
$ podman run --rm -it  --mount type=tmpfs,destination=/dev/shm  alpine
/ #
```

[the remark this answers is not in the document]
Comment (89luca89) [reply distrobox-create]: I'm not sure I can support WSL at this point
If we need to do so many exceptions (which I cannot test on Windows at all) to support WSL2, that basic distrobox functionality (Like sharing mount points) is nerfed, I prefer to keep distrobox not compatible with it

Probably using docker (rooftul) as backend maybe is better?

**Comment by 89luca89** [comment]
@OhMyMndy 

From what I understand, by default the user in the actions is already running podman in rootless mode

**Pull request closed without merging**